### DISCUSSION TITLE
Z2T Home Assistant MQTT auto discovery

### DISCUSSION THREAD

**Discussion by urechi** [arendst/Tasmota#16609] [Ideas]
I believe one of the missing features from Z2T is Home Assistant MQTT auto discovery. This is probably one of the main reasons people will choose Zigbee2MQTT or ZHA instead of Z2T. I am stubborn and chose Z2T but It took me a while to manually configure over 200+ entities from 80+ Zigbee devices. Not everyone is willing and has the time to do this so I believe this auto discovery feature will just attract more people to use Tasmota for their ZigBee hubs. Along with the recent commit from @s-hadinger that allows people to map the rare device attributes much easier, I think Z2T should get increased usage.

**Comment by barbudor** [comment]
I believe one of the missing feature of HA, or more exactly HaTasmota plug-in, is the ability to support dynamically additions of items.
I already had a discutions with the author regarding BLE sensors who is insisting in getting all discovered devices aggregated in a single discovery message. Every time a new device is discovered, it should be added to the discovery message that is then resent.
While it would work with less than ten devices, the size of the message and the associated memory it would require would quickly go over the ESP capacity.

If HaTasmota is OK to receive dynamically 1 discovery message per BLE/Zb sensor discovered, this could be done.

As of today no core developers of Tasmota is using HA at home...
I'm personally using Node-red and mostly everything in my infra is based on auto-detection from published topics without the need of dedicated discovery message. So that's possible too from the HA authors.

**Reply by urechi** [reply]
While node-red is a nice piece of software, HA is got his advantages in comparison for home automation. Don’t get me wrong, i like node red, unusually i’ve implement it at work as a security orchestration, automation and response (SOAR) tool and works great for the tasks. That being said, at home i still prefer HA over node red and that’s mostly because of the native app offered by HA.
I also see your point so thanks for sharing your thoughts with us.

**Reply by barbudor** [reply]
I'm not saying NR is better than HA
Just that what I have been able to do in NR, could be done in HA
Don't push back the complexity of what can be handled on a Rpi/PC to the ESP.

**Reply by s-hadinger** [reply]
I couldn't agree more. Resources in the ESP are **very** limited, while a Raspberry Pi has roughly 1000x more resources. Unfortunately I don't use HA. If you can find a volunteer to work on the HA plugin...

**Reply by sfromis** [reply]
In principle, someone finding an elegant way of doing it might be able to create the mapping to work like "classic" hardware for auto discovery, using the Berry language on ESP32. It should be feasible for such code to construct a map of "recently seen" payloads, to format it for HA. Still takes quite a bit of working out details for the many different device types, with a high risk of the code needing to be adapted to work with some devices. Not a project for me :grimacing:

**Reply by urechi** [reply]
Home Assistant automation could be also used to republish MQTT messages from the Tasmota `SENSOR` topic to HA MQTT discovery format, like in this example [here](https://community.home-assistant.io/t/mqtt-demultiplexer-automatic-sensor-discovery-for-tasmota2zigbee/237900/5). I remember I've tried this a few years back and it worked decently.
